### ISSUE TITLE
Allow editors to "back out" of the review, copyediting or production stages

### ISSUE THREAD

**Issue by rcgillis** [pkp/pkp-lib#2890]
A client has requested the capacity to be able to go back to a previous stage in the editorial workflow, in cases where mistakes have been made and work still needs to be done in a preceding stage (e.g. sent to copyediting too early and work still needs to be done in the review stage). 

Considerations to be taken into account: 

- How does it effect the editorial history on a submission
- Effect on statistics
- Existing reviews, review files, drafts, etc. preserved within the record 
- Having a feature that like an close/open stage feature, which preserved the editorial history and all associated assignments but allowed an editor to close down a stage if it's been opened too early

**Update 2022-04-04:** The ability to back out of most stages was added with #7265. However, work still needs to be done to support backing out of a review stage (eg - closing review assignments, notifying reviewers, updating review round status).

**Update 2022-08-10:** Work on backing out of review rounds is in progress. PRs:
[pkp-lib](https://github.com/pkp/pkp-lib) --> [pkp/pkp-lib#2890 Allow editors to back out of the review, copyediting or production stages](https://github.com/pkp/pkp-lib/pull/8031)
[ojs](https://github.com/pkp/ojs) --> [pkp/pkp-lib#2890 Allow editors to back out of the review, copyediting or production stages](https://github.com/pkp/ojs/pull/3438)
[omp](https://github.com/pkp/omp) --> [pkp/pkp-lib#2890 Allow editors to back out of the review, copyediting or production stages](https://github.com/pkp/omp/pull/1145)

**Comment by NateWr** [comment]
A few more questions to gather requirements:

- Is the goal to completely shut down the stage that was opened too early? ie - Must no one be able to access it or perform actions on it?
- Is the goal to better track progress, so that the "correct" stage is shown in submission lists?
- Is there a point-of-no-return for a stage? Should a review stage with active reviews be allowed to "back out"?

Depending on the requirements, a "set current stage" feature that doesn't effect the workflow, but simply resets the active stage that appears in lists, might be sufficient.

**Comment by Mariaros** [comment]
To answer the questions above: 

-  Yes, the goal is to shut it down, and that no one can access it. 
-  Yes, it would be best if the “correct” stage showed in submission list, and that the wrong stage is deleted. 
-  No, we would not like there to be an opening to back out, once the process of for example reviewing has started, that is the point of no return. 
"Depending on the requirements, a "set current stage" feature that doesn't effect the workflow, but simply resets the active stage that appears in lists, might be sufficient."
 Yes, this would be sufficient
A situation where we need to shut down a review-round for example, would be  where the editor has pressed "Resubmit for review" instead of "request revision".

**Comment by NateWr** [comment]
> A situation where we need to shut down a review-round for example, would be where the editor has pressed "Resubmit for review" instead of "request revision".

Can you describe this use-case in more detail? Neither of those actions ("Resubmit for Review" or "Request Revisions") moves the submission to the next stage, so I'm not sure how a "close stage" feature would help.

**Comment by Mariaros** [comment]
yes, sorry. I see now that it is only when they actively choose "New Review round" that the submission moves to the next stage. The "close stage" feature would help to revert the action, where they have chosen "New review round" but they did not intend to do this. In order to fix the mistake, we currently have to move both reviewers and comments to round 2.

**Comment by NateWr** [comment]
@Mariaros Are the review rounds the only place that you're having trouble with this?

**Comment by Mariaros** [comment]
No, we have also experienced that editors press "send to copyediting" before the MS is ready. We would also appriciate a way to revert this action if possible. In general, we find it somewhat difficult to reverse actions in the platform.

**Comment by NateWr** [comment]
:+1: Thanks, that's helpful feedback!

**Comment by hkamareddineiiit** [comment]
Hello,

I think this is an important feature to work on because sometimes it is done by mistake and to ensure the integrity of the process, a step back is needed.

Thanks!
huda

**Comment by librariam** [comment]
+1 from PKP|PS

**Comment by NateWr** [comment]
I expect that at least partial support for this will be included as part of work for #7265. It will be possible to go forward or backward a stage, but maybe not to close a review round.

**Comment by AlejandraCasas** [comment]
+1 from PKP|PS

**Comment by NateWr** [comment]
This is related to https://github.com/pkp/pkp-lib/issues/3585.

**Comment by asmecher** [comment]
@NateWr, has this been mooted by the new decision toolset?

**Comment by NateWr** [comment]
We can back out of every stage except the review stage now. The review stage is the one people want the most, though. :)

It shouldn't be too difficult to add a decision to back out of the review stage, it just needs to handle a few extra things (closing review assignments, notifying reviewers, changing review round status).

I've added a note to the original issue description.

**Comment by NateWr** [comment]
> It shouldn't be too difficult to add a decision to back out of the review stage, it just needs to handle a few extra things (closing review assignments, notifying reviewers, changing review round status).

In addition to these things, we may also need to consider decisions that move a submission into the review stage. They may need to check if a review round already exists and, if so, re-open it.

**Comment by touhidurabir** [comment]
I can see that reverting back to the `Review Stage` ability already added now in the `main`. As I test with the latest pull form `main` branch, I can confirm the it work properly when reverting back to review stage from. All review rounds are there , both the last running one and any previous one . Also the last review round get activated immediately after reverting back to review stage with proper assigned reviewer with their last status intact when it was moved to next stage(e.g.  copy editing stage) . @NateWr may be in one of your last few commits you already fixed it or perhaps I am missing something here , can you please confirm ? 

The only missing piece that I can see is that when reverting back to review stage , the assigned reviewers in a open review round who still haven't responded or did not submit any review after accepting it, do not get any mail notification which I think is necessary to notify them .

**Comment by touhidurabir** [comment]
So when back to `Review Round`, we have 2 cases to consider 

1. ability to get back to `Submission` directly regardless of which review round it's currently at as there may be just a single review round as multiple .
2. go back to previous review round one by one if there are multiple and in the last review round only have the option to get back to submission .

In case of No: 1 seems easy as that will only need to revert back to `submission` and if want to go through `Review` again , just need to open the last review round again.

But for multiple review rounds it become bit difficult as we need to have option to back and forth between the review rounds also that is if there are 5 review rounds and revert back to round 3 , then have the option to move to review round 4 also if editor wants to . 

So in that process in, when reverting back to review, we need to show both of the options that is `Back to Submission` with `Back to Review Round ##` and `Forward to Review Round ##` when there are multiple exist. 

I guess above approach would be the ideal implementation for us even though it's bit lengthy , right @NateWr ?

**Comment by NateWr** [comment]
Yeah, this is why this feature is a bit complicated. Backing out of a review round can leave a lot of things in an uncertain state, like review assignments. We don't want someone in review round 3 to be able to send something directly back to the submission stage.

In particular, the following question/comments above:

> Is there a point-of-no-return for a stage? Should a review stage with active reviews be allowed to "back out"?

> No, we would not like there to be an opening to back out, once the process of for example reviewing has started, that is the point of no return.

> The "close stage" feature would help to revert the action, where they have chosen "New review round" but they did not intend to do this.

Based on those comments, I think the most important use case is when someone opens a review round by accident. So I think we should only permit someone to back out of a review round there are no active reviews (ie - a review is assigned but does not have the "cancelled" status). 

That will support almost all of the cases, but keep things simple on our end. We can simply delete the review round and pretend it never happened.

**Comment by touhidurabir** [comment]
> Based on those comments, I think the most important use case is when someone opens a review round by accident. So I think we should only permit someone to back out of a review round if there are no active reviews (ie - a review is assigned but does not have the "cancelled" status).

@NateWr  I am bit confused why the `cancelled` status of the `review_assignments` table is so important and based on that we want to make the decision . An active review round can be in several state 

1. Multiple reviewers already assigned and no response form any of them yet . 
2. Multiple reviewers already assigned and few have accepted and few have declined . 
3. Multiple reviewers already assigned and few have already submitted reviews .
4. Multiple reviewers already assigned and few have already submitted reviews and some some of those are accepted and some got declined by the editor . 

In those above cases do we want to allow editor to back out form the `current review round` or a `previous round` if there is multiple round or if `only one is running` , then back to `Submission` ? 

Also what happens if one get to back out from current round to previous round or to submission ? Do we need to have that backed up round in the list so that they can also go forward to that ? Removing those does not seems like a good idea . 

Also the `cancelled` status only get to set(that is set to 1) only when an editor or journal manager cancelled an assigned reviewer . so why that is so significant only ? 

So basically I am bit confused based on what criteria we will allow an editor to back our from a review round ? For example , I think we should not allow them to back out if `there is already one or more reviews submitted by reviewers` . Also may be association with assigned reviewers , not sure though .

**Comment by touhidurabir** [comment]
@NateWr for the `OMP` , if a submission pass down to `Copy Editing` stage form the `Internal Review` only(that is there was no external review stage) , it shows the options `Back To Submission` . Should it not show the option as `Back To Internal Review` ? is this intentional or a mistake ? Also it seems that the back out does not work when testing in the `main` branch . 

But if it pass down from the `External Review` stage , it shows the options `Back To Review` which works properly and moved it back to the last round of external review and set stage properly .

**Comment by NateWr** [comment]
@touhidurabir I probably just cut corners here and didn't bother implementing the additional decisions in OMP. If something is there but not working, that should get fixed. If something is not there at all, we might wait to see if there's real demand for it in OMP before implementing it. That said, if you think you can implement it quickly, then go for it.

**Comment by touhidurabir** [comment]
@NateWr ok, I going to implement it also as this also includes in my list of decisions to complete to get this issue done .

**Comment by touhidurabir** [comment]
PRs 
[pkp-lib](https://github.com/pkp/pkp-lib) --> [pkp/pkp-lib#2890 Allow editors to back out of the review, copyediting or production stages](https://github.com/pkp/pkp-lib/pull/8031)
[ojs](https://github.com/pkp/ojs) --> [pkp/pkp-lib#2890 Allow editors to back out of the review, copyediting or production stages](https://github.com/pkp/ojs/pull/3438)
[omp](https://github.com/pkp/omp) --> [pkp/pkp-lib#2890 Allow editors to back out of the review, copyediting or production stages](https://github.com/pkp/omp/pull/1145)

@asmecher @NateWr please review it. @Vitaliy-1 please take a look at the `OJS` and `OMP` pulls to see how the new email templates added for the upgrade and if any changes required.

The decision implementation follows as 

### OJS 
1. Back To Previous External Review Stage : This is when multiple External Review Round is going on and this decision allow editor to back out from one active review round to previous review round.
2. Back To Submission from External Review Stage : This is when the submission in the review round stage and there is only one review round running
3. Back to External Review from Copy Editing Stage : This is when the submission in the copy editing stage and there is one or more review round available in the review stage. 
4. Delete Empty External Review Round : This is when there is multiple review round running and the last review round has no reviewer assigned to it yet . 

### OMP
1. Back to Previous Round (External/Internal Round) Stage : This is when multiple External Review Round is going on and this decision allow editor to back out from one active review round to previous review round.
2. Back To Submission from External Review Stage : This is when the submission in the external review round stage and there is only one review round running .
3. Back to Submission from Internal Review Stage : This is when the submission in the internal review round stage and there is only one review round running .
4. Back to Internal Review Stage from External review Stage : This is when the submission in the external review round stage and there is only one review round running .
5. Back to External review Stage from Copy Editing Stage : This is when the submission in the copy editing stage and there is one or more review round available in the external review stage.
6. Back to Internal review Stage from Copy Editing Stage : This is when the submission in the copy editing stage and there is one or more review round available in the internal review stage.
7. Delete Empty Review(Internal/External) Round : This is when there is multiple review round running and the last review round has no reviewer assigned to it yet .

**Comment by NateWr** [comment]
Great work @touhidurabir! I did a code review of pkp-lib and ojs, and left some comments on those PRs. I did a little bit of testing in practice, but I'll take a closer look in the next round of review. In addition to the comments on the PRs, I had the following thoughts:

1. Instead of having different decisions for back to submission, back one round, etc, let's make one decision to "Cancel Review Round". When that decision is recorded, if there is only one round in the current stage, it will go back to the previous stage. If there is more than one review, it will stay in the current stage. I think this will make things a little bit clearer to the editor, because it's not clear from the "Back to Submission" button that the review round will be deleted. And I think it will reduce the number of different decisions you need.

2. The list of editorial decision buttons will need to be updated when a review assignment is confirmed. Otherwise you will end up with outdated buttons. Here are reproduction steps:
- Send a submission to review
- See "Back to Submission" in the buttons on the right
- Add a reviewer
- Open that reviewer's actions and click "Review Details"
- Select a recommendation and click "Confirm"
- See the "Back to Submission" is still there, but when you click on it it leads to a 404.
Fixing this will be a little challenging. I think we have something in the code already that forces an update of the editorial actions panel. You'll want to look for use of something like `setGlobalEvent` and `bindGlobal` or something like that. If you can't find it, let me know and I'll have a look.

3. No need to do it yet, but when we're getting close to merge, can you squash your commits so that there are 1 or a few meaningful commits? It helps us manage the git history when we're looking back years later trying to figure out when something changed.

**Comment by Vitaliy-1** [comment]
@touhidurabir, email template installation looks good! Sorry, it's probably a headache to rebase the email stuff while the code is constantly changing.

**Comment by touhidurabir** [comment]
@NateWr changes based on code review are pushed . Please take a look . I have made modifications for most the code review that you have pointed out except just few as I think a bit more digging and debate is required for those cases . 

> The list of editorial decision buttons will need to be updated when a review assignment is confirmed. Otherwise you will end up with outdated buttons

I am bit struggling on this one but I think I can figure it out . If I can't, I will ask for your help. 

> can you squash your commits so that there are 1 or a few meaningful commits

I have squashed the commits into fewer commits . if needed, I will squash it more into 1/2 commits when we are ready to merge .

**Comment by NateWr** [comment]
Looks good @touhidurabir! I did some manual testing in OJS and OMP and everything worked well. I ran into a bit of an issue in OMP, but after looking through the code review again I think i forgot to apply the migrations.

I've left a few comments here and there for some things. But I think mostly we just have the one conversation remaining about whether or not to use `canRetract`. Otherwise, we're almost there! :+1:

**Comment by touhidurabir** [comment]
@NateWr second batch of code modification based on your second batch of code review is pushed. There are few conflicts that I need to revolve so once we decided to marge , I will fix those . please review the changes .

**Comment by touhidurabir** [comment]
@NateWr I have added the new `BackFromXXXXXX` decision classed and associated modifications . Haven't removed the previous decision classes code yet as we need to first come to a conclusion about which patterns we finally decide to go with . Once we come to a final decision, I will do detailed clean up and remove all the dead codes . Please review it .

**Comment by NateWr** [comment]
Looks good. I've left a couple of comments but I think the approach looks good. Happy to look at a final version for review.

I noticed that some unrelated code has snuck into the PR (see PKPDoiHandler and MailablerHandler). I suspect his is because of the `.inc.php` renamed to `.php`. A rebase should sort it out...

**Comment by touhidurabir** [comment]
@NateWr dead codes and old decisions classes, email templates and unused locales clean up done . Please review it .

**Comment by NateWr** [comment]
That's great, @touhidurabir! I did a final round of testing on the decisions and everything works great. I left some comments on the PRs, but they are just changes to the language that we're using in messages/emails.

I also prepared a couple PRs with some suggestions for reducing the number of traits that are in play. I feel it's just a little bit too complex with 9 traits, some of which depend on each other. So I made some suggestions that would cut that down to six, and reduce the overall number of decisions. You can take these changes or leave them, depending on what you think. It's up to you and I'm happy to merge as-is. If you like them, you can merge them into your branches:

https://github.com/touhidurabir/pkp-lib/pull/1
https://github.com/touhidurabir/omp/pull/1

I also found a bug in my own decision code. If you can merge this fix into your OJS branch that should sort it out:

https://github.com/touhidurabir/ojs/pull/1

**Comment by touhidurabir** [comment]
@NateWr I have merged all three of your PRs . there was few bug lest there so I have fixed those and pushed the changes . please review it .

**Comment by NateWr** [comment]
Great, @touhidurabir! It looks like we're just about ready for merge. I was able to get the editorial decisions to reload when reviewer details are changed. I also made a couple more minor changes to the language of emails, and moved some shared decision constants into pkp-lib:

https://github.com/touhidurabir/pkp-lib/pull/2
https://github.com/touhidurabir/ojs/pull/2
https://github.com/touhidurabir/omp/pull/2

The tests look like they are mostly passing. The test failures look unrelated to your code. But there are enough of them that I'd like to run them one more time. Once you've merged those PRs, go ahead and kick off the tests again. We'll see if any of the test failures are consistent and, if not, I'll go ahead and merge tomorrow. :+1: 

(ps - I edited the first comment in the issue to include links to your PRs since GitHub was burying the middle comments.)

**Comment by touhidurabir** [comment]
@NateWr All PRs have been merged and I also run some local test, no issue . Now if the tests passes properly , I think we are good to merge it .

**Comment by touhidurabir** [comment]
@asmecher I have squashed the commits and rebased it to fix the conflicts . If the upgrade tests passes, it's good to merge with main .

**Comment by asmecher** [comment]
Thanks, @touhidurabir! I rushed the tests (I broke them yesterday) but ran them locally and will monitor the Travis jobs. Congratulations on this work!

**Comment by NateWr** [comment]
@emmauhl this will effect documentation for 3.4. Short summary: it is now possible to "back out" of every stage of the workflow, including a review round that's been started. Once a reviewer has completed a review, though, we prevent cancelling that round.

**Comment by ajnyga** [comment]
Thanks for this work! This is one of those new features that reduce the amount of support tickets from journals in a very concrete way. I usually get at least a couple of requests to cancel a review round that has been accidentally created and so far I've had to edit the database directly.
